### PR TITLE
8325495: C2: implement optimization for series of Add of unique value

### PR DESCRIPTION
**Pull request by tabjy**
This pull request resolves [JDK-8325495](https://bugs.openjdk.org/browse/JDK-8325495) by converting series of additions of the same operand into multiplications. I.e., `a + a + ... + a + a + a => n*a`.

As an added benefit, it also converts `C * a + a` into `(C+1) * a` and `a << C + a` into `(2^C + 1) * a` (with respect to constant `C`). This is actually a side effect of IGVN being iterative: at converting the `i`-th addition, the previous `i-1` additions would have already been optimized to multiplication (and thus, further into bit shifts and additions/subtractions if possible).  

Some notable examples of this transformation include:
- `a + a + a` => `a*3` => `(a<<1) + a`
- `a + a + a + a` => `a*4` => `a<<2`
- `a*3 + a` => `a*4` => `a<<2`
- `(a << 1) + a + a` => `a*2 + a + a` => `a*3 + a` => `a*4 => a<<2`

See included IR unit tests for more.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325495](https://bugs.openjdk.org/browse/JDK-8325495): C2: implement optimization for series of Add of unique value (**Enhancement** - P4)

### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) Review applies to [af6f8084](https://git.openjdk.org/jdk/pull/20754/files/af6f808472bc34eedbd095b3f1e2c5957f0c1694)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20754/head:pull/20754` \
`$ git checkout pull/20754`

Update a local copy of the PR: \
`$ git checkout pull/20754` \
`$ git pull https://git.openjdk.org/jdk.git pull/20754/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20754`

View PR using the GUI difftool: \
`$ git pr show -t 20754`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20754.diff">https://git.openjdk.org/jdk/pull/20754.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20754#issuecomment-2316377986)

### CODE REVIEW

**Comment by tabjy** [comment]
Also, please feel free to suggest better naming for `SerialAdditionCanonicalization` (test) and `AddNode::find_repeated_operand_in_chained_addition`.

**Comment by tabjy** [comment]
> The existing test compiler/c2/TestLargeTreeOfSubNodes.java times out...

I'm currently investigating this. The test is unrolled and produces a chained subtraction of ~180 nodes and causing performance regression. This will take a little bit of time.

**Comment by dean-long** [comment]
Can we redistrute the repeated terms and turn `a + x + a + y + a` into `a*3 + x + y`?

**Comment by tabjy** [comment]
@dean-long No it cannot, although I agree this is a fairly tempting feature to have. Maybe this can be looked into with another issue, but I believe re-ordering addition is out-of-scope for this one.

**Comment by tabjy** [comment]
> If this is your intention, then please ignore this message.

Yes, this is my intention.

--- 

My previous approach of identifying optimized `Mul->shift + add/sub` (e.g., `a*6` becomes `(a<<1) + (a<<2)` by `MulNode::Ideal()`) was inherently flawed. I was solely determining this with the number of terms. It is not reliable. In the `TestLargeTreeOfSubNodes` example, it replaces already optimized Mul nodes and a new Mul node and repeats the process, causing performance regression (and timeouts).

The new approach matches the exact patterns of optimized `MulNode`s. Additionally, a recursion depth limit of 5 (a rather arbitrary number) is in effect during *iterative* GVN to mitigate the risk of exhausting resources. Untransformed nodes are added to the worklist and will be eventually transformed.

Please note, in the case of `TestLargeTreeOfSubNodes` with flags mentioned above, the compilation is skipped without a large enough `-XX:MaxLabelRootDepth`. This is the same behaviour as the current master. 

Please re-review once GHA is confirmed passing. Thanks!

**Comment by chhagedorn** [comment]
> Please note, in the case of TestLargeTreeOfSubNodes with flags mentioned above, the compilation is skipped without a large enough -XX:MaxLabelRootDepth. This is the same behaviour as the current master.

Have you found out why this is the case? I thought that the original fix wanted to fix the problem of running out of nodes.

I gave your patch another spin. We still see various failures and timeouts. For example:

`compiler/intrinsics/sha/TestDigest.java` times out with various flag combinations (for example `-server -Xmixed`). Here is the stack at the timeout:

```
Thread 7 (Thread 0x7fc808490700 (LWP 22433)):
#0  0x00007fc80d648051 in Node::find_integer_type(BasicType) const () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#1  0x00007fc80c793214 in AddNode::extract_base_operand_from_serial_additions(PhaseGVN*, Node*, Node**, int) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#2  0x00007fc80c79306c in AddNode::extract_base_operand_from_serial_additions(PhaseGVN*, Node*, Node**, int) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
...
#90 0x00007fc80c79306c in AddNode::extract_base_operand_from_serial_additions(PhaseGVN*, Node*, Node**, int) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#91 0x00007fc80c793082 in AddNode::extract_base_operand_from_serial_additions(PhaseGVN*, Node*, Node**, int) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#92 0x00007fc80c79306c in AddNode::extract_base_operand_from_serial_additions(PhaseGVN*, Node*, Node**, int) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#93 0x00007fc80c793351 in AddNode::convert_serial_additions(PhaseGVN*, bool, BasicType) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#94 0x00007fc80c7937c5 in AddNode::IdealIL(PhaseGVN*, bool, BasicType) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#95 0x00007fc80d73ea47 in PhaseGVN::transform(Node*) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#96 0x00007fc80d715296 in Parse::do_one_bytecode() () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#97 0x00007fc80d7029ca in Parse::do_one_block() () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#98 0x00007fc80d703e86 in Parse::do_all_blocks() () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
#99 0x00007fc80d707738 in Parse::Parse(JVMState*, ciMethod*, float) () from /opt/mach5/mesos/work_dir/jib-master/install/2024-09-17-0714032.christian.hagedorn.jdk-test/linux-x64-debug.jdk/jdk-24/fastdebug/lib/server/libjvm.so
...
```

I'm also seeing the live node limit assert with test
```
applications/ctw/modules/java_desktop.java
```
and flags:
```
-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation -Djava.awt.headless=true	
```
Assert hit:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/opt/mach5/mesos/work_dir/slaves/a20696e7-ae7d-4d37-8e9c-83f99ef002cb-S27847/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/118d2aab-d4da-49da-972c-f7b6fd5b5eee/runs/4c2607c3-6f9f-49ea-8d16-a0fb44652951/workspace/open/src/hotspot/share/opto/node.cpp:79), pid=305278, tid=305301
#  assert(C->live_nodes() <= C->max_node_limit()) failed: Live Node limit exceeded limit
#
# JRE version: Java(TM) SE Runtime Environment (24.0) (fastdebug build 24-internal-2024-09-17-0714032.christian.hagedorn.jdk-test)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 24-internal-2024-09-17-0714032.christian.hagedorn.jdk-test, mixed mode, sharing, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x146c697]  Node::verify_construction()+0x1a7
```

File: src/hotspot/share/opto/addnode.cpp [hunk 1]
```diff
@@ -395,9 +395,141 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  if (node->is_Mul()
+      && node->Opcode() != Op_AndI && node->Opcode() != Op_AndL // AndNode extends MulNode for some reason
+      && (node->in(1)->is_Con() || node->in(2)->is_Con())) {
+    Node* const_node = node->in(1)->is_Con() ? node->in(1) : node->in(2);
+    Node* operand_node = node->in(1)->is_Con() ? node->in(2) : node->in(1);
+    BasicType bt = phase->type(const_node)->basic_type();
+
+    if (bt == T_INT || bt == T_LONG) { // const could potentially be void type
+      Node* mul_base;
+      jlong multiplier = extract_base_operand_from_serial_additions(phase, operand_node, &mul_base, depth_limit - 1);
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:490]: Do you need to recurse at all here?

File: src/hotspot/share/opto/addnode.cpp [hunk 2]
```diff
@@ -395,9 +395,141 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  // 3. input is an already-optimized multiplication, i.e., a*6 => (a<<1) + (a<<2)
+  // 4. is simple additions not worth transforming, i.e., a + a
+  if (base == nullptr
+      || base == this
+      || is_optimized_multiplication(this, base)
+      || (base == in(1) && base == in(2))) {
+    return nullptr;
+  }
+
+  Node* con = (bt == T_INT) ? (Node*) phase->intcon((jint) factor) : (Node*) phase->longcon(factor);
```
[the remark this answers is not in the document]
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp:427]: I disagree: `integercon()` internally uses `checked_cast<jint>(l)` to make prevent information loss during type conversion and asserts at runtime if the value is larger what a `jint` can hold. However, such an information loss is intended for integer arithmetic overflows. (e.g., `Integer.MAX_VALUE * a + a` is extracted to `(Integer.MAX_VALUE + 1) * a`. Here we want `(long) (Integer.MAX_VALUE) + 1` to overflow to `(int) Integer.MIN_VALUE`).

If I were to use `integercon()`, the best I could do is `intgercon(bt == T_INT ? (jint) factor : factor)` which is rather pointless.

File: src/hotspot/share/opto/addnode.cpp (hunk 1 above)
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp]: I believe so. Consider the case `(a + a) * 3`. Recurse here allows us to extract `a` and factor `2 * 3 => 6`

File: src/hotspot/share/opto/addnode.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rwestrel) [reply src/hotspot/share/opto/addnode.cpp:427]: Good catch.
It makes sense to add a comment about that in the source code.
Do you have a test case for that corner case?

File: src/hotspot/share/opto/addnode.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/addnode.cpp]: That case would be better handled in 2 steps, I think:
`a+a` into `a*2` with a `AddNode` transformation
`(a*2)*3` into `a*6` with a `MulNode` (or `LShift`) transformation. Can you check if it already exists?

File: src/hotspot/share/opto/addnode.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp]: Added comments.

> Do you have a test case for that corner case?

Yes: `TestSerialAdditions::mulAndAddToOverflow`

**Comment by tabjy** [comment]
@chhagedorn 

> Have you found out why this is the case? I thought that the original fix which added TestLargeTreeOfSubNodes wanted to fix the problem of running out of nodes.

I think this is intended. The original fix does prevents running out of nodes during optimization. However, the compilation skipped during code gen, not IGVN. The optimization, after all, correctly unrolls into a large number of nodes which default value of 1100 for `MaxLabelRootDepth` prevents compilation.

> We still see various failures and timeouts. For example: `compiler/intrinsics/sha/TestDigest.java`

This happened because `node->is_Add()` is deceivingly true when `node` is an `Xor[I/L]Node`. I adopted Roland's suggestions, and it's now comparing opcodes directly. 

> I'm also seeing the live node limit assert with test `applications/ctw/modules/java_desktop.java`

There was a problem detecting optimized power-of-2 multiplication where the base term itself is also a `LShiftNode`. It's fixed now.

---

After some discussions with Roland, I decided to not to use recurssion at all to avoid risk of exhausting resource. The latest version matches the exact patterns of multiplications optimized into power-of-2 additions. This should be much safer. It passes all `TEST="tier1 hotspot_compiler tier1_compiler tier2_compiler tier3_compiler tier1_compiler_not_xcomp"` tests. Please give it another run. Thank you very much!

@rwestrel Could you please give it a quick look as well? Thanks!

File: src/hotspot/share/opto/addnode.cpp (hunk 1 above)
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp]: Done. The new version doesn't use recurssions at all

File: src/hotspot/share/opto/addnode.cpp [hunk 3]
```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  //     - AddNode(LShiftNode(a, CON), LShiftNode(a, CON)/a)
+  //     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+  // given that lhs is different from rhs
+  if (lhs == rhs) {
+    return false;
+  }
+
+  // swap LShiftNode to lhs for easier matching
+  if (!lhs->is_LShift()) {
+    Node* tmp = lhs;
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:562]: You can use swap() here.

File: src/hotspot/share/opto/addnode.cpp [hunk 4]
```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+  int in1_op = in1->Opcode();
+
+  bool matched = false;
+  jlong multiplier;
+
+  // Convert (a + a) + a to 3 * a
+  // Look for LHS pattern: AddNode(a, a)
+  if (in1_op == Op_Add(bt) && in1->in(1) == in1->in(2)) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:422]: It seems each of the if blocks in this method could be its own method that returns true and `multiplier` (passed by reference, I suppose) if pattern matching succeeds.

File: src/hotspot/share/opto/addnode.cpp [hunk 5]
```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+    Node* lhs_base = lhs->is_LShift() ? lhs->in(1) : lhs;
+    Node* rhs_base = rhs->is_LShift() ? rhs->in(1) : rhs;
+
+    Node* nodes[] = {lhs, rhs};
+    Node* bases[] = {lhs_base, rhs_base};
+    jlong multipliers[] = {1, 1};
+
+    // AddNode(LShiftNode(a, CON1), LShiftNode(a, CON2)/a)
+    // AddNode(LShiftNode(a, CON1)/a, LShiftNode(a, CON2))
+    for (int i = 0; i < 2; i++) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:487]: I wouldn't use a loop here. I would put the loop body into its own method and call it twice, once with `lhs`, `lhs_base` as arguments, once with `rhs`, `rhs_base`.

File: src/hotspot/share/opto/addnode.cpp [hunk 6]
```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (in1_op == Op_LShift(bt) && in1->in(2)->is_Con()) {
+    Node* base = in1->in(1);
+    if (base != in2) {
+      return nullptr;
+    }
+
+    Node* con = in1->in(2);
+    BasicType con_bt = phase->type(con)->basic_type();
+    if (con_bt == T_VOID) { // const could potentially be void type
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:442]: Does that happen when con is top? In that case I think it's better to use `con->is_top()`. Or maybe `find_integer_as_long` if there's a value we know for sure that con can't take (it can't be negative, right?).

File: src/hotspot/share/opto/addnode.cpp [hunk 7]
```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  }
+
+  multiplier++; // + 1 for current addition
+  Node* con = (bt == T_INT)
+              ? (Node*) phase->intcon((jint) multiplier) // intentional type narrowing to allow overflow at max_jint
+              : (Node*) phase->longcon(multiplier);
+  Node* mul = MulNode::make(con, in2, bt);
+
+  PhaseIterGVN* igvn = phase->is_IterGVN();
+  if (igvn != nullptr) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:540]: Why do you need that?
I think it's fine to return a new node from Ideal.

File: src/hotspot/share/opto/addnode.cpp [hunk 8]
```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Convert a + a + ... + a into a*n
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (is_optimized_multiplication()) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:409]: What would happen if that `if` is removed? Would matching in the rest of the method accept something that's not a valid pattern?

Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp:409]: It wouldn't accept invalid patterns. It still makes (semantically) correct transformation but there is a risk it can't progress. The `if` here is to safeguard repeatedly adding terms of an already optimized mul node. Otherwise there would be an eventual timeout/live node limit situation. (E.g., `3 * a` => `(a << 2) + a` => `3 * a` => ...)

File: src/hotspot/share/opto/addnode.cpp (hunk 3 above)
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp:562]: Good point. That is more idiomatic c++.

File: src/hotspot/share/opto/addnode.cpp (hunk 6 above)
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp:442]: Yes it is better. Will use`is_top()` instead.

> if there's a value we know for sure that con can't take (it can't be negative, right?)

I thought about `find_integer_as_long(-1)`, but it could be anything, even negative. This transformation accounts for example `a*-43 + a` => `a*-42`.

File: src/hotspot/share/opto/addnode.cpp (hunk 7 above)
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp]: You are right. This is leftover code from last version. Removed.

File: src/hotspot/share/opto/addnode.cpp (hunk 5 above)
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp]: I refactored even further to combine checking for optimized `mul`s and extracting multipliers to use the same logic. This code is now obsolete.

File: src/hotspot/share/opto/addnode.cpp (hunk 4 above)
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp]: Refactored to do so. Thanks for the input!

File: src/hotspot/share/opto/addnode.cpp [hunk 9]
```diff
@@ -395,9 +395,178 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Convert a + a + ... + a into a*n
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (find_power_of_two_addition_pattern(this, bt, nullptr) != nullptr) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:409]: Can you a comment that explain the need for this (what you replied in the PR comment essentially)?

File: src/hotspot/share/opto/addnode.cpp [hunk 10]
```diff
@@ -395,9 +395,178 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  // Look for patterns:
+  //     - AddNode(LShiftNode(a, CON), LShiftNode(a, CON)/a)
+  //     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+  // given that lhs is different from rhs
+  if (n->Opcode() == Op_Add(bt) && n->in(1) != n->in(2)) {
+    Node* lhs = n->in(1);
+    Node* rhs = n->in(2);
+
+    // swap LShiftNode to lhs for easier matching
+    if (!lhs->is_LShift()) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:498]: Can you use `Op_LShift(bt)` here?

File: src/hotspot/share/opto/addnode.cpp [hunk 11]
```diff
@@ -395,9 +395,178 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+      swap(lhs, rhs);
+    }
+
+    // AddNode(LShiftNode(a, CON), *)?
+    if (!lhs->is_LShift() || !lhs->in(2)->is_Con()) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:503]: Same here.

File: src/hotspot/share/opto/addnode.cpp [hunk 12]
```diff
@@ -395,9 +395,178 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+    if (lhs->in(1) == rhs) {
+      if (multiplier != nullptr) {
+        *multiplier = lhs_multiplier + 1;
+      }
+
+      return rhs;
+    }
+
+    // AddNode(LShiftNode(a, CON), LShiftNode(a, CON2))?
+    if (rhs->is_LShift() && lhs->in(1) == rhs->in(1) && rhs->in(2)->is_Con()) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:527]: same here.

File: src/hotspot/share/opto/addnode.cpp [hunk 13]
```diff
@@ -395,9 +395,178 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+    return nullptr;
+  }
+
+  return nullptr;
+}
+
+// Match `(a << CON) - a`, extract `a` and `(1 << CON) - 1`
+Node* AddNode::find_power_of_two_subtraction_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  // Look for pattern: SubNode(LShiftNode(a, CON), a)
+  if (n->Opcode() == Op_Sub(bt) && n->in(1)->is_LShift() && n->in(1)->in(2)->is_Con()) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/addnode.cpp:549]: same here.

**Review by rwestrel — COMMENTED** [review]
Thanks for making the changes. It's easier to follow the various steps the way it is now.

**Comment by rwestrel** [comment]
@chhagedorn would you mind running the latest version patch through testing?

**Comment by TobiHartmann** [comment]
Testing all passed.

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for running tests

**Review by rwestrel — APPROVED** [review]

File: src/hotspot/share/opto/addnode.hpp [hunk 14]
```diff
@@ -42,6 +42,15 @@ typedef const Pair<Node*, jint> ConstAddOperands;
 // by virtual functions.
 class AddNode : public Node {
   virtual uint hash() const;
+
+private:
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.hpp:46]: Can be removed since these methods are already private by default here since it's a `class` and not a `struct`.
```suggestion
```

File: src/hotspot/share/opto/addnode.cpp [hunk 15]
```diff
@@ -395,9 +395,181 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+      return nullptr;
+    }
+
+    Node* con = lshift->in(2);
+    if (con->is_top()) {
+      return nullptr;
+    }
+
+    // We can't simply return the lshift node even if ((a << CON) - a) + a cancels out. Ideal() must return a new node.
+    *multiplier = ((jlong) 1 << con->get_int()) - 1;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.cpp:567]: Can't this be an `Identity()` transformation where you can return existing nodes?

File: test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    // ----- integer tests -----
+    @Test
+    @Arguments(values = {Argument.RANDOM_EACH})
+    @IR(counts = { IRNode.ADD_I, "1" })
+    @IR(failOn = {IRNode.LSHIFT_I})
+    private static void addTo2(int a) {
+        int sum = a + a; // Simple additions like a + a should be kept as-is
+        verifyResult(a, 2, sum);
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:64]: Generally, we should move all verification code out of the `@Test` methods to avoid side effects and worrying about whether the result checking is now compiled or not (we must ensure that the result checking code is interpreted to catch wrong executions with miscompiled code).

I suggest the following (not tested):

Introduce a `@Run` method, which is never compiled, for your `@Test` methods. You can still call methods from there but then you should ensure that they are not compiled either with `@DontCompile`:
```
    static final Random RANDOM = Utils.getRandomInstance();

    ...

    @DontCompile
    private static void verifyResult(int base, int factor, int observed) { ... }

    ...

    @Test
    @IR(counts = {IRNode.ADD_I, "1", IRNode.LSHIFT_I, "1"})
    private static int addTo3(int a) {
        return a + a + a; // a*3 => (a<<1) + a
    }

    @Run(test = "addTo3") 
    void runAddTo3() {
        int a = RANDOM.nextInt();
        int result = addTo3(a);
        verifyResult(a, 3, result);
    }
```
Since the tests are all very similar and require the same setup and verification, you could even go a step further and provide a single shared `@Run` method which is possible:
```
    @Test
    @IR(counts = {IRNode.ADD_I, "1", IRNode.LSHIFT_I, "1"})
    private static int addTo3(int a) {
        return a + a + a; // a*3 => (a<<1) + a
    }

    @Test
    @IR(failOn = IRNode.ADD_I)
    @IR(counts = {IRNode.LSHIFT_I, "1"})
    private static int addTo4(int a) {
        return a + a + a + a; // a*4 => a<<2
    }

    @Run(test = {"addTo3", "addTo4"}) // List all @Test methods here and make sure you call all of them below.
    void runTests() {
        int a = RANDOM.nextInt();
        verifyResult(a, 3, addTo3(a));
        verifyResult(a, 4, addTo4(a));
    }

```
This also allows you to run with some more edge case values like `a == 0` or `a == min_int` etc. which gives us even some more confidence.

File: test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+        if (expected != observed) {
+            throw new AssertionError("Expected " + expected + " but got " + observed);
+        }
+    }
+
+    // ----- integer tests -----
+    @Test
+    @Arguments(values = {Argument.RANDOM_EACH})
+    @IR(counts = { IRNode.ADD_I, "1" })
+    @IR(failOn = {IRNode.LSHIFT_I})
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:61]: Generally, for single strings, you can remove the braces:
```suggestion
    @IR(failOn = IRNode.LSHIFT_I)
```

File: test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+    private static void addTo2(int a) {
+        int sum = a + a; // Simple additions like a + a should be kept as-is
+        verifyResult(a, 2, sum);
+    }
+
+    @Test
+    @Arguments(values = {Argument.RANDOM_EACH})
+    @IR(counts = { IRNode.ADD_I, "1" })
+    @IR(counts = {IRNode.LSHIFT_I, "1"})
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:70]: Generally, you can merge these together:
```suggestion
    @IR(counts = {IRNode.ADD_I, "1", IRNode.LSHIFT_I, "1"})
```

**Review by chhagedorn — CHANGES_REQUESTED** [review]
Good updates, it's now easy to follow the logic and understand the code. I have some more comments/suggestions.

File: src/hotspot/share/opto/addnode.cpp [hunk 19]
```diff
@@ -395,9 +395,181 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+  // Look for pattern: AddNode(a, a)
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    *multiplier = 2;
+    return n->in(1);
+  }
+
+  return nullptr;
+}
+
+// Match `a << CON`, extract `a` and `1 << CON`
```
[the remark this answers is not in the document]
Comment (tabjy) [reply src/hotspot/share/opto/addnode.cpp:446]: Updated comments. Thanks for the suggestion!

File: test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java (hunk 16 above)
Comment (tabjy) [reply test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:64]: Thanks for the idea. Converted to custom `@Run` methods and test with `a = 0, 1, min, max, rand`

File: src/hotspot/share/opto/addnode.cpp [hunk 20]
```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
...
+                ? (Node*) phase->intcon((jint) multiplier) // intentional type narrowing to allow overflow at max_jint
+                : (Node*) phase->longcon(multiplier);
+    return MulNode::make(con, in2, bt);
+  }
+
+  return nullptr;
+}
+
+// Try to match `a + a`. On success, return `a` and set `2` as `multiplier`.
+// The method matches `n` to for pattern: AddNode(a, a).
```
Comment (chhagedorn) [inline src/hotspot/share/opto/addnode.cpp:439]:
```suggestion
// The method matches `n` for pattern: AddNode(a, a).
```

File: test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+            "addTo8", //
+            "addTo16", //
+            "addAndShiftTo16", //
+            "addTo42", //
+            "mulAndAddTo42", //
+            "mulAndAddToMax", //
+            "mulAndAddToOverflow", //
+            "mulAndAddToZero", //
+            "mulAndAddToMinus1", //
+            "mulAndAddToMinus42" //
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java:65]: Why did you add the trailing `//`?

File: src/hotspot/share/opto/addnode.cpp (hunk 15 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/addnode.cpp]: That's great!

**Review by chhagedorn — APPROVED** [review]
Thanks for the updates! Good conversion of the tests. 

ll give this another spinning in our testing.

**Comment by chhagedorn** [comment]
Testing looked good.

File: test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java (hunk 21 above)
Comment (tabjy) [reply test/hotspot/jtreg/compiler/c2/TestSerialAdditions.java]: Those are added to prevent formatter from collapsing these lines to one. I've gone ahead to remove them.

**Comment by tabjy** [comment]
> Testing looked good.

Thank you @chhagedorn. Could please grant an approval once again (after updates only to comments) so we can merge this. Thanks!

**Review by chhagedorn — APPROVED** [review]
Still good, thanks!
